test/EndToEndTests/Tests/Core/Microsoft.OData.Core.E2E.Tests/WriteJsonPayloadTests/WritePayloadTests.cs

```diff
@@ -0,0 +1,1056 @@
+                outputWithoutModel = rex.Replace(outputWithoutModel, "");
+            }
+
+            Assert.Equal(outputWithModel, outputWithoutModel);
+        }
+
+        [Theory]
+        [InlineData(MimeTypeODataParameterMinimalMetadata, true, false, false)]
+        [InlineData(MimeTypeODataParameterNoMetadata, false, false, false)]
+        [InlineData(MimeTypeODataParameterFullMetadata, true, false, false)]
+        public async Task WritingAFeedContainingActionsAndDerivedTypes_ShouldMatchExpectedPayload(string mimeType, bool hasMetadata, bool hasODataType, bool hasNoODataType)
+        {
+            var settings = new ODataMessageWriterSettings
+            {
+                BaseUri = _baseUri,
+                ODataUri = new ODataUri() { ServiceRoot = _baseUri },
+                EnableMessageStreamDisposal = false
+            };
+
+            string outputWithModel, outputWithoutModel;
+
+            var responseMessageWithModel = new TestStreamResponseMessage(new MemoryStream());
+            responseMessageWithModel.SetHeader("Content-Type", mimeType);
+
+            var personType = _model.FindDeclaredType(NameSpacePrefix + "Person") as IEdmEntityType;
+            var personSet = _model.EntityContainer.FindEntitySet("People");
+
+            using (var messageWriter = new ODataMessageWriter(responseMessageWithModel, settings, _model))
+            {
+                var odataWriter = await messageWriter.CreateODataResourceSetWriterAsync(personSet, personType);
+                outputWithModel = await WriteAndVerifyPersonFeedAsync(responseMessageWithModel, odataWriter, false, mimeType);
+            }
+
+            var responseMessageWithoutModel = new TestStreamResponseMessage(new MemoryStream());
+            responseMessageWithoutModel.SetHeader("Content-Type", mimeType);
+
+            using (var messageWriter = new ODataMessageWriter(responseMessageWithoutModel, settings))
+            {
+                var odataWriter = await messageWriter.CreateODataResourceSetWriterAsync();
+                outputWithoutModel = await WriteAndVerifyPersonFeedAsync(responseMessageWithoutModel, odataWriter, false, mimeType);
+            }
+
+            Assert.Equal(outputWithModel, outputWithoutModel);
+
+            if (hasMetadata)
+            {
+                Assert.Contains(_baseUri + "$metadata#People\"", outputWithoutModel);
+            }
+
+            if (hasODataType)
+            {
+                Assert.True(outputWithoutModel.Contains("{\"@odata.type\":\"" + "#" + NameSpacePrefix + "Employee\","), "odata.type Employee");
+                Assert.True(outputWithoutModel.Contains("{\"@odata.type\":\"" + "#" + NameSpacePrefix + "SpecialEmployee\","), "odata.type SpecialEmployee");
+            }
+
+            if (hasNoODataType)
+            {
+                Assert.False(outputWithoutModel.Contains("{\"@odata.type\":\"" + "#" + NameSpacePrefix + "Person\","), "odata.type Person");
+            }

```

Comment:
If you're going to parameterize the test, is it possible to put parameters that contain actual values you're going to assert against, instead for conditions for whether or not you should execute the assertions.

For example, maybe you could have a parameter that contains the actual metadata information that should appear in the payload, then assert `Assert.True(outputWithoutModel.StartsWith(payloadStart)` instead of using if statements.